### ISSUE TITLE
Fixing BatchNorm_LinearLayer

### ISSUE THREAD

**Issue by AErbsloeh** [es-ude/elastic-ai.creator#396]
Transforming BatchNorm Linear from Creator into Model does not work (error in training process, values are looking random and difference between real and what we get is huge)

**Comment by glencoe** [comment]
Can you go into more detail wrt. error in training process? What error? How are you finishing training to generate code if there is an error?

I think we already know that the difference is due to the batch norm parameters in full res being merged with the fixed point convolution weights.

I suggest to 

1. quantize the batchnorm as well, in a first step using the same (total, frac)-bits as the convolution. 
2. Raise a warning in case the merged result is not representable with (total_bits, frac_bits) fixed point.

**Comment by Sillobear** [comment]
**Previous Method:**  
1. Train a Linear layer followed by a BatchNorm layer.  
2. Fold the BatchNorm weights into the Linear layer's weights.  
3. Generate a design for the Linear layer.  

**Issues:**  
During training, the BatchNorm layer is not quantized, which can lead to quantization errors when folding the weights later. The training process appears normal.  

To investigate further, I added a second Linear layer:  
![image](https://github.com/user-attachments/assets/0440ca46-cdda-4d56-90d9-006fcbb5f415)  

During evaluation, we fold the weights of the BatchNorm and the first Linear layer into the second Linear layer:  
![image](https://github.com/user-attachments/assets/6042105a-47fc-48de-8474-37604bd0b927)  

This produces noticeably different results, indicating that folding the weights is not a viable option for our quantized BatchNormedLinear:  
![image](https://github.com/user-attachments/assets/c34edd3c-cccb-4f33-98c4-050fcaf45dca)

**Comment by Sillobear** [comment]
### Current calculation when folding:
$$\frac{(W^TX + B) + E[X]}{\sqrt{Var[x] - \epsilon}} \ast \gamma + \beta = \gamma \ast \frac{W}{\sqrt{Var[X] - \epsilon}}X + (\gamma * \frac{B - E[X]}{\sqrt{Var[x] - \epsilon}} + \beta)$$

Problem: $Var[X], E[X], \epsilon, \gamma, \beta$ aren't quantized and therefore quantization errors arise when folding the functions.

**Comment by glencoe** [comment]
> **Previous Method:**
> 
>     1. Train a Linear layer followed by a BatchNorm layer.
> 
>     2. Fold the BatchNorm weights into the Linear layer's weights.
> 
>     3. Generate a design for the Linear layer.
> 

@Sillobear : please use code fence blocks instead of images in the future, that is way easier to process and work with here on github
